src/main/java/org/checkerframework/checker/crypto/CryptoAnnotatedTypeFactory.java

```diff
@@ -1,57 +1,87 @@
     }
 
-    private static final class CryptoQualifierHierarchy extends GraphQualifierHierarchy {
+    private final class CryptoQualifierHierarchy extends ElementQualifierHierarchy {
 
-        CryptoQualifierHierarchy(MultiGraphFactory factory, AnnotationMirror bottom) {
-            super(factory, bottom);
+        /**
+         * Creates a CryptoQualifierHierarchy from the given classes.
+         *
+         * @param qualifierClasses classes of annotations that are the qualifiers for this hierarchy
+         * @param elements element utils
+         */
+        public CryptoQualifierHierarchy(
+                Collection<Class<? extends Annotation>> qualifierClasses, Elements elements) {
+            super(qualifierClasses, elements);
         }
 
         @Override
         public boolean isSubtype(final AnnotationMirror subtype, final AnnotationMirror supertype) {
-            if (AnnotationUtils.areSameByClass(supertype, UnknownAlgorithmOrProvider.class)
-                    || AnnotationUtils.areSameByClass(subtype, Bottom.class)) {
+            if (AnnotationUtils.areSameByName(supertype, UNKNOWNALGORITHMORPROVIDER)
+                    || AnnotationUtils.areSameByName(subtype, BOTTOM)) {
                 return true;
-            } else if (AnnotationUtils.areSameByClass(subtype, UnknownAlgorithmOrProvider.class)
-                    || AnnotationUtils.areSameByClass(supertype, Bottom.class)) {
+            } else if (AnnotationUtils.areSameByName(subtype, UNKNOWNALGORITHMORPROVIDER)
+                    || AnnotationUtils.areSameByName(supertype, BOTTOM)) {
                 return false;
-            } else if (AnnotationUtils.areSameByClass(subtype, AllowedAlgorithms.class)
-                    && AnnotationUtils.areSameByClass(supertype, AllowedAlgorithms.class)) {
+            } else if (AnnotationUtils.areSameByName(subtype, ALLOWEDALGORITHMS)
+                    && AnnotationUtils.areSameByName(supertype, ALLOWEDALGORITHMS)) {
                 return compareAllowedAlgorithmOrProviderTypes(subtype, supertype);
-            } else if (AnnotationUtils.areSameByClass(subtype, AllowedProviders.class)
-                    && AnnotationUtils.areSameByClass(supertype, AllowedProviders.class)) {
+            } else if (AnnotationUtils.areSameByName(subtype, ALLOWPROVIDERS)
+                    && AnnotationUtils.areSameByName(supertype, ALLOWPROVIDERS)) {
                 return compareAllowedAlgorithmOrProviderTypes(subtype, supertype);
             } else {
                 return false;
             }
         }
 
+        @Override
+        public AnnotationMirror greatestLowerBound(AnnotationMirror a1, AnnotationMirror a2) {
+            return CryptoAnnotatedTypeFactory.this.BOTTOM;
+        }
+
+        @Override
+        public AnnotationMirror leastUpperBound(AnnotationMirror a1, AnnotationMirror a2) {
+            return CryptoAnnotatedTypeFactory.this.UNKNOWNALGORITHMORPROVIDER;
```

Comment:
I'll merge this PR now, but this looks like a very coarse over-approximation. What is a1 and a2 are the same annotation? Why return Unknown? Same for the GLB. Please think about example test cases and prepare a separate PR that fixes GLB and LUB.